openapi-core/src/main/java/io/ballerina/openapi/core/generators/type/generators/AllOfRecordTypeGenerator.java

```diff
@@ -205,6 +223,49 @@ private ImmutablePair<List<Node>, List<Schema<?>>> generateAllOfRecordFields(Lis
         return ImmutablePair.of(recordFieldList, validSchemas);
     }
 
+    private void handleCommonRequiredFields(List<String> requiredFields, Schema refSchema, List<Node> recordFieldList)
+            throws OASTypeGenException {
+        try {
+            Map<String, Schema<?>> properties = getPropertiesFromRefSchema(refSchema);
+            List<String> required = refSchema.getRequired();
+            if (Objects.nonNull(required)) {
+                requiredFields.removeAll(required);
+            }
+            Set<Map.Entry<String, Schema<?>>> fieldProperties = new HashSet<>();
+            for (Map.Entry<String, Schema<?>> property : properties.entrySet()) {
+                if (requiredFields.contains(property.getKey())) {
+                    fieldProperties.add(property);
+                }
+            }
+            recordFieldList.addAll(addRecordFields(requiredFields, fieldProperties, typeName));
+        } catch (BallerinaOpenApiException e) {
+            throw new OASTypeGenException(e.getMessage());
+        }
+    }
+
+    private static Map<String, Schema<?>> getPropertiesFromRefSchema(Schema schema) throws InvalidReferenceException {
+        Map<String, Schema<?>> properties = schema.getProperties();
+        if (Objects.isNull(properties)) {
+            properties = new HashMap<>();
+        }
+        String refSchema = schema.get$ref();
+        if (Objects.nonNull(refSchema)) {
+            String extractedSchemaName = GeneratorUtils.extractReferenceType(refSchema);
+            OpenAPI openAPI = GeneratorMetaData.getInstance().getOpenAPI();
+            Schema<?> refSchemaObj = openAPI.getComponents().getSchemas().get(extractedSchemaName);
+            if (Objects.nonNull(refSchemaObj)) {
+                properties.putAll(getPropertiesFromRefSchema(refSchemaObj));
+            }
+        }
+        List<Schema<?>> allOfSchemas = schema.getAllOf();
+        if (Objects.nonNull(allOfSchemas)) {
+            for (Schema<?> allOfSchema : allOfSchemas) {
+                properties.putAll(getPropertiesFromRefSchema(allOfSchema));
+            }
+        }
+        return properties;
+    }
+
```

Comment:
Shall we add a test for this?

Reply:
Added via [91ce874](https://github.com/ballerina-platform/openapi-tools/pull/1712/commits/91ce8740930bbdbc2e3588e2569243e8207401f1)